## Functions as Arguments - Input Arguments

In [18]:
def hello(name):
    print("Hello!", name)

In [19]:
hello("Bruce")

Hello! Bruce


In [20]:
hello

<function __main__.hello(name)>

In [21]:
greet = hello

In [22]:
greet("Tom")

Hello! Tom


In [23]:
calculate_length = len

In [24]:
calculate_length("SomeString")

10

In [25]:
calculate_length == len

True

In [26]:
calculate_length([3, 5, 6, 8, 1, 2])

6

In [27]:
import math

In [28]:
def area_circle_fn(radius):
    
    return math.pi * radius * radius

In [29]:
def perimeter_circle_fn(radius):
    
    return 2 * math.pi * radius

In [30]:
def diameter_circle_fn(radius):
    
    return 2 * radius

In [31]:
def calculate_for_circle(radius, fn): ## Function as an input argument
    return fn(radius)

In [32]:
calculate_for_circle(10, diameter_circle_fn)

20

In [33]:
calculate_for_circle(10, perimeter_circle_fn)

62.83185307179586

In [34]:
calculate_for_circle(10, area_circle_fn)

314.1592653589793

In [35]:
calculate_for_circle(20, perimeter_circle_fn)

125.66370614359172

## Functions as Arguments - Variable Arguments

In [26]:
def calculate(*args, fn):
    return fn(*args)

In [28]:
calculate(10, fn=diameter_circle_fn)

20

In [29]:
calculate(10, fn=perimeter_circle_fn)

62.83185307179586

In [30]:
calculate(10, fn=area_circle_fn)

314.1592653589793

In [31]:
def area_rectangle_fn(length, breadth):
    
    return length * breadth

In [32]:
def perimeter_rectangle_fn(length, breadth):
    
    return 2 * (length + breadth)

In [33]:
calculate(20, 40, fn=area_rectangle_fn)

800

In [36]:
calculate(20, 40, fn=perimeter_rectangle_fn)

120

In [35]:
calculate(10, fn=area_rectangle_fn)

TypeError: perimeter_rectangle_fn() missing 1 required positional argument: 'breadth'

## Functions as Return Values

In [40]:
def add(a, b):
    return a + b

def sub(a, b):
    return a - b

def mul(a, b):
    return a * b

def div(a, b):
    return a / b

In [41]:
def get_function(operator='+'):
    
    if operator == "+":
        return add ## these return values are functions

    if operator == "-":
        return sub
    
    if operator == "*":
        return mul 
    
    if operator == "/":
        return div 

In [42]:
func = get_function()

func

<function __main__.add(a, b)>

In [43]:
func(3, 4)

7

In [44]:
func = get_function('*')

func

<function __main__.mul(a, b)>

In [45]:
func(3, 4)

12

In [46]:
calc_dictionary = {
    '+': add,  ## these values are functions
    '-': sub,
    '*': mul,
    '/': div
}

In [47]:
func = calc_dictionary['/']

func

<function __main__.div(a, b)>

In [48]:
func(12, 4)

3.0

In [49]:
func = calc_dictionary['*']

func

<function __main__.mul(a, b)>

In [50]:
func(10, 20)

200

In [51]:
calc_dictionary['*'](100, 3)

300

In [52]:
calc_dictionary['+'](100, 3)

103

# Video: Functions as Arguments - Input Arguments 

Objectives
pass functions as input arguments to other functions
[Video description begins] Topic title: Functions as Arguments - Input Arguments. Your host for this session is Janani Ravi. [Video description ends]

### We've worked a lot with functions in Python, and you should know that functions in Python are first-class citizens.

[Video description begins] A Jupyter notebook titled FirstClassFunctions is displayed on the screen. It has a menu bar with the following options: File, Edit, View, etc. A tool bar is present with the following buttons: Save, add, cut, run, etc. A set of code cells is present in the notebook. [Video description ends]

Functions are first-class, in that you can do with functions everything that you can do with other types of data whether they are primitive types, such as string, number, booleans, and so on, or complex data types, such as lists, dictionaries, sets, etc. What does it mean for functions to be first-class citizens? That's exactly what we'll see here in this demo. Let's consider a custom function that I have defined here. This function is named hello and it takes in one input argument, name, and I've printed out Hello and name.

[Video description begins] There are two code lines in code cell 1. Line 1 is: def hello(name):. Line 2 is: print ("Hello!", name). [Video description ends]

So if invoke this hello function and pass in Bruce, Bruce will be assigned to the name input argument and the output of this function will be Hello Bruce.

[Video description begins] Code cell 2 is: hello ("Bruce"). Its output is: Hello! Bruce. [Video description ends]

Now, strings and other data types are all objects in Python, whether they are complex data types or primitive data types. Hello, which is a function is also an object in Python. So, if you simply type the output of hello, you can see that it's a function object.

[Video description begins] Code cell 3 is: hello. Its output is: <function _main_.hello( name)>. [Video description ends]

#### Everything is an object in Python including functions. And this function object is stored somewhere in Python's memory, and is called when we invoke the function. What does it mean when we say functions are first-class citizens? This means that functions can be assigned to variables just like all other data types in Python. For example, you can create a new variable called greet and assign it this function, hello.

[Video description begins] Code cell 4 is: greet = hello. [Video description ends]

What we're essentially doing with this assignment is that the variable greet refers to the same function object in memory. So if you call greet and invoke it as a function and pass in the input argument, Tom, you'll get Hello Tom.

[Video description begins] Code cell 5 is: greet("Tom"). Its output is: Hello! Tom. [Video description ends]

You can see that the function object is now available in the greet variable. And if you use the greet variable to invoke the function, you're essentially invoking the hello function. It's not just custom functions, user defined functions that you have defined that you can assign to variables. 

#### In fact, you can reference a built-in function object such as the len function object and assign that to a variable as well. Here I've assigned it to the calculate_length variable.

[Video description begins] Code cell 6 is: calculate_length= len. [Video description ends]

Now, I can invoke the calculate_length function. Let's say I pass in a string as an input argument and the built-in len function will be invoked on the string. The length of the string is 10.

[Video description begins] Code cell 7 is: calculate_length("SomeString"). [Video description ends]

Let's use the equality comparison operator to see whether the calculate_length function object is the same as the built-in len function object. calculate_length == len will give us True, indicating that they refer to the same function.

[Video description begins] Code cell 8 is: calculate_length == len. [Video description ends]

Let's invoke the calculate_length function on this list here, a list of numbers, and you can see that it returns 6. What you're invoking is simply the built-in len function.

[Video description begins] Code cell 9 is: calculate_length ([3, 5, 6, 8, 1, 2]). [Video description ends]

What else is the significance of functions as first-class citizens? That's exactly what we'll see in the rest of this demo. I'm going to go ahead and import the math module so I can use some useful functionality from within it.

[Video description begins] Code cell 10 is: import math. [Video description ends]

Next, I'm going to define a function which calculates the area of a circle. I've called this function the area_circle function. And its input argument, the only input argument is the radius of the circle. If you remember your math from back in high school, the area of a function is given by the formula, pi r square and that's exactly the formula that I've used in the body of the function to calculate the area. And the area is what this function returns.

[Video description begins] There are two code lines in code cell 11. Line 1 is: def area_circle_fn(radius):. Line 2 is: return math.pi * radius * radius. [Video description ends]

I'm going to define another function here called perimeter_circle function which takes as its input argument the radius of the circle. And then uses the formula 2 pi r in order to calculate the perimeter of the circle.

[Video description begins] There are two code lines in code cell 12. Line 1 is: def perimeter_circle_fn(radius):. Line 2 is: return 2 * math.pi * radius. [Video description ends]

And now, I'm going to define one more custom function called diameter_circle function. Once again, the input argument is exactly 1, that is the radius of the circle. And 2 multiplied by the radius gives us the diameter and that's what this function returns.

[Video description begins] There are two code lines in code cell 13. Line 1 is: def diameter_circle_fn(radius):. Line 2 is: return 2 * radius. [Video description ends]

#### And here is one more reason for functions to be considered as first-class citizens in Python. Functions can be passed in as an input argument to another function, just like any other data type in Python. Observe this function that I have defined here, calculate_for_circle. What do I calculate for a circle? Well, that depends on what input argument you specify.

#### This function takes in two input arguments. The first of these is the radius of the circle and the second input argument is a function object. And this function object defines the calculation that we want to perform on the circle whether it's calculating the area, perimeter, or the diameter. Once you accept this function as an input argument, you can simply invoke the function passed in. And pass in the radius as an input argument to that function.

#### [Video description begins] There are two code lines in code cell 14. Line 1 is: def calculate_for_circle(radius, fn ):. Line 2 is: return fn(radius). [Video description ends]

All we do within the body of the calculate_for_circle function is to invoke fn, fn is the function that we passed in. The input argument to fn is radius and whatever value this invocation of the fn function returns is what we return from the calculate_for_circle function. What I just described here is actually pretty exciting and mind bending. Let's see this in action.

I'm going to invoke the calculate_for_circle function. The radius of my circle is 10 and the operation that I want to perform or what I want to calculate for this circle is its diameter. All I need to do here is pass in the diameter_circle function as a second input argument which will be assigned to the fn argument in calculate_for_circle.

[Video description begins] Code cell 15 is: calculate_for_circle(10, diameter_circle_fn ). Screenshots of code cells 13 and 14 appear. [Video description ends]

Within the body of the calculate_for_circle function, the diameter_circle function will be invoked and the radius will be passed in. And the result that this will give us is the diameter of the circle which is 20. What if I now want to know the perimeter of the circle? Well, you can call calculate_for_ circle, pass in the radius. And the second input argument will be the function which will actually perform the perimeter calculation.

[Video description begins] Code cell 16 is: calculate_for_circle(10, perimeter_circle_fn ). A screenshot of code cell 12 appears. [Video description ends]

And the result of this function invocation is the perimeter for this particular circle which happens to be around 62.83. Observe how we've used the idea of functions being first-class citizens to really customize this calculate_for_circle function. By passing in another function which performs the actual calculation.

[Video description begins] Code cell 17 is: calculate_for_circle(10, area_circle_fn ). A screenshot of code cell 11 appears . [Video description ends]

Here is how we calculate the area of this circle with a radius of 10. And the result displays the area of the circle with the radius 10, which is around 314.15. Now, if you specify a different value for the radius, the function will operate on this new value for radius that you've specified. Here I've invoked the calculate_for_circle function with radius 20 and I want to calculate the perimeter. The perimeter_circle function will operate on the radius value that we passed in, which in this particular case happens to be 20.

[Video description begins] Code cell 18 is: calculate_for_circle(20, perimeter_circle_fn ). [Video description ends]

And the result is the area of the circle with radius 20, which is around 125.66.

## Functions as Arguments - Variable Arguments 

Objectives
work with functions that accept any number of arguments
[Video description begins] Topic title: Functions as Arguments - Variable Arguments. Your host for this session is Janani Ravi. [Video description ends]

The calculate for circle function which accepted a function as an input argument from the last video was so-called because it could only perform calculations for circles. Why is that? Because it assumes that the only input argument that you need to perform calculations for area, perimeter, diameter and so on, is the single radius input argument.

[Video description begins] A Jupyter notebook titled FirstClassFunctions is displayed on the screen. It has a menu bar with the following options: File, Edit, View, etc. A set of code cells is present in the center pane. At the bottom right corner, a screenshot is displayed. It contains two lines of code. Line 1 is: def calculate_for_circle(radius, fn):. Line 2 is: return fn(radius). [Video description ends]

#### If you think about it like that, the calculate for circle function is rather limiting. You won't be able to use the calculate_for_circle function if you want to perform area or perimeter calculations for say a rectangle where you'll need a length as well as the breadth of the rectangle. So clearly, instead of forcing this function to take in just one input argument, say the radius, we need to make it more generic. How can we do this? Well, that is why the *args are variable length input arguments come in.

[Video description begins] Two code lines appear in code cell 19. Line 1 is: def calculate(*args, fn):. Line 2 is: return fn(*args). [Video description ends]

Here is a more generic calculate function, where the first input argument is *args and the second input argument is where we take in a function object. The *args will allow us to specify variable numbers of input arguments. And the function object that we passed in can then be invoked using these variable length arguments. Observe that within the body function here, we invoke the function object and pass in as an input argument to this function object *args. Invoking the fn function object with *args might seem a little confusing to you at first, but don't be.

It's actually quite simple. The presence of the asterisk when we invoke the fn object tells Python to unpack the args tuple. So remember, args inside the function has the variable length arguments in the tuple format. When you use *args while invoking the fn object, the args tuple will be unpacked into individual elements. If this seemed a little confusing, let's understand this using an example.

#### Let's say the args tuple contained two arguments, two values, 10, 2. Args is a tuple, *args will unpack this tuple so that the function object fn is invoked with 10, 2. The fn function object will not receive a tuple as an input, it will receive the individual elements of the tuple. By simply using variable length input arguments, we've made our calculate function more generic, more powerful. This can now work with rectangles, triangles, circles, you name it. Let's now use this calculate function to calculate the diameter of a circle.

[Video description begins] The following code appears in code cell 20: calculate(10, fn=diameter_circle_fn). [Video description ends]

Remember that the first input argument is a variable length argument. Now in order to calculate the diameter of a circle, all you need is the radius. So we pass in 10 as the radius of the circle. And the function object that is to be invoked for the calculation is the diameter_circle_function. This works perfectly well, the diameter of the circle with radius 10 is 20. I'll use the calculate function once again, this time to calculate the perimeter of the circle with radius 10. Make sure the fn object is perimeter_circle_function.

[Video description begins] The following code appears in code cell 21: calculate(10, fn=perimeter_circle_fn). [Video description ends]

I'm using keyword arguments here instead of positional argument, so that it's very clear what the function object is all about.

[Video description begins] The following value appears below code cell 21: 62.83185307179586. [Video description ends]

Here is the perimeter of the circle with radius 10. Let's confirm that calculating the area of the circle also works.

[Video description begins] The following code appears in code cell 22: calculate(10, fn=area_circle_fn). [Video description ends]

The radius is 10 and the function that we want to apply is area_circle_function. The area of this circle is roughly 314.159 which is correct. I just mentioned a little while earlier that this calculate function with variable length arguments that is *args allows us to perform calculations on other kinds of shapes as well, which require more than one input argument, any number of input arguments. Here is a function to calculate the area of a rectangle.

[Video description begins] Two code lines appear in code cell 23. Line 1 is: def area_rectangle_fn(length, breadth):. Line 2 is: return length * breadth. [Video description ends]

It takes as its input the length and breadth of the rectangle, length multiplied by breadth is the area, and that's what this function returns. I'm going to define one more custom function here to calculate the perimeter of a rectangle, perimeter rectangle function is the name it takes in the length and the breadth of the rectangle.

[Video description begins] Two code lines appear in code cell 24. Line 1 is: def perimeter_rectangle_fn(length, breadth):. Line 2 is: return 2 * (length + breadth). [Video description ends]

2 multiplied by length plus breadth gives us the perimeter. Now let's perform area calculation using the calculate function.

#### [Video description begins] The following code appears in code cell 25: calculate(20, 40, fn=area_rectangle_fn). At the bottom right corner, a screenshot appears. It displays the code present in code cell 19. [Video description ends]

Remember, calculating the area of a rectangle requires two input arguments, 20, 40 will be assigned to the variable length argument args and the function that we want to invoke to perform this calculation is the area rectangle function. The area rectangle function requires two input arguments, the length and breadth of the rectangle. Length will be assigned 20, breadth will be assigned 40, giving us an area of this rectangle 800. The area_rectangle_function will multiply 20 by 40, giving us the result 800. What if we want to calculate the perimeter of this rectangle?

[Video description begins] The following code appears in code cell 26: calculate(20, 40, fn=perimeter_rectangle_fn). [Video description ends]

Well, you can use the calculate function once again. The variable length arguments are 20 and 40, which correspond to the length of the rectangle and breadth of the rectangle, respectively. And the function which actually performs the perimeter calculation is perimeter_rectangle_function. 2 multiplied by length plus breadth will give us 120. 2 multiplied by 20 + 40. Now, what if you don't specify the right number of input arguments for the function object that you want to apply to perform the calculation, like in this example here?

[Video description begins] The following code appears in code cell 27: calculate(10, fn=area_rectangle_fn). A screenshot appears. It displays the code present in code cell 23. [Video description ends]

The area_rectangle_function requires the length and breadth of the rectangle. But we've specified only the value for one input argument here, that is 10. We haven't specified values for both length and breadth which means when we try to invoke the area_rectangle_function on 10, there will be an error and that's exactly what you see here on screen.

#### [Video description begins] The output for code cell 27 displays a TypeError. It reads: area_rectangle_fn() missing 1 required positional argument: 'breadth'. [Video description ends]


When area_rectangle_function is invoked on 10, there is one missing required positional argument, the value for breadth. We only have the value for length specified here, which is why this function throws an error.

##  Functions as Return Values 

Objectives
return functions from other functions
[Video description begins] Topic title: Functions as Return Values. Your host for this session is Janani Ravi. [Video description ends]

Let's continue our discussion of first class functions.

[Video description begins] A Jupyter notebook titled FirstClassFunctions is displayed on the screen. It has a menu bar with the following options: File, Edit, View, etc. Eight code lines are present in code cell 28. Line 1 is: def add(a, b):. Line 2 is: return a + b. Line 3 is: def sub(a, b):. Line 4 is: return a - b. Line 5 is: def mul(a, b):. Line 6 is: return a * b. Line 7 is: def div(a, b):. Line 8 is: return a / b. [Video description ends]

We haven't yet explored and discovered all of the cool things that you can do with functions in Python. I'm going to build a simple calculator here using the idea that functions are first class citizens in Python. I've defined four simple arithmetic operations using different functions. The add function takes in two input arguments a and b, and returns a+b.

Sub takes a and b and returns a minus b. The mul function takes in a and b, and returns a multiplied by b. And the div function takes in two input arguments a and b, and returns a divided by b. As I set up my little calculator function, you'll see why, once again, functions can be considered to be first class citizens in Python. This is because functions can be return values from functions.

[Video description begins] Nine code lines appear in code cell 29. Line 1 is: def get_function(operator='+'):. Line 2 is: if operator == "+":. Line 3 is: return add. Line 4 is: if operator == "-":. Line 5 is: return sub. Line 6 is: if operator == "*":. Line 7 is: return mul. Line 8 is: if operator == "/":. Line 9 is: return div. [Video description ends]

#### You can return a function from within another function, just like you can with other data types. I've defined a custom function here called get function which takes as its input argument an operator. It has a default value where the operator is +. Based on what you specify as the operator input argument, get function will check to see what the operator is, and return the right function corresponding to that operator.

[Video description begins] A screenshot of code cell 28 appears. [Video description ends]

If it's the plus operator that you specified as an input argument, we'll return the add function. If the operator is minus, we'll return the sub function, if it's multiplication, we'll return the mul function and if it's the division operator, we'll return the div function. Observe how all of the return statements in this get function return function objects. Based on the kind of operation that you want to perform, you can invoke the get_function function.

[Video description begins] Two lines of code appear in code cell 30. Line 1 is: func = get_function(). Line 2 is: func. [Video description ends]

If you specify no input argument, this will return the addition function. The default value for the operator is +, after all. If you print out func to screen, you can see that this refers to the add function.

[Video description begins] The following output appears: <function _main_.add(a, b)>. [Video description ends]

The add function object is what is returned from get function when you specify no input argument. Once you have a reference in the func variable to this add function object, you can use it to add two numbers.

[Video description begins] The following code appears in code cell 31: func(3, 4). [Video description ends]

Here, I'm invoking the func function on three and four, and you can see that the result is the addition of three and four, 3+4=7. Let's access another function here using the get function object.

[Video description begins] Two lines of code appear in code cell 32. Line 1 is: func = get_function('*'). Line 2 is: func. [Video description ends]

I pass in the asterisk operator that is the multiplication sign, the function that is returned will be the mul function.

[Video description begins] The following output appears: <function _main_.mul(a, b)>. [Video description ends]

Now that we have a reference to the mul function object, in the func variable, we can invoke this function pass in three and four, we get the multiplication result equal to 12.

[Video description begins] The following code appears in code cell 33: func(3, 4). The output is 12. [Video description ends]

Before we complete our discussion on first class functions in Python, let's discuss one more way in which functions are first class citizens.

[Video description begins] Six lines of code appear in code cell 34. Line 1 is: calc_dictionary = {. Line 2 is: '+': add,. Line 3 is: '-': sub,. Line 4 is: '*': mul,. Line 5 is: '/': div. Line 6 is: }. [Video description ends]

#### Function objects can be elements in a list or values in a Python dictionary, just like any other data type. I have set up a calc dictionary here, and the keys in this dictionary are arithmetic operators, plus, minus, multiplication, and the division signs. Observe the values in this dictionary, all of the values are function objects. The value associated with the plus key is the add function, with the minus key is the sub function, with the multiplication key is the mul function, and with the division key is the div function. Once you have a dictionary with all of the function objects, you can look up these function objects by key.

[Video description begins] Two lines of code appear in code cell 35. Line 1 is: func = calc_dictionary['/']. Line 2 is: func. [Video description ends]

Here is how you can access the division operation from our calc_dictionary. And I've assigned this division function to the func variable.

[Video description begins] The following output appears: <function _main_.div(a, b)>. [Video description ends]

If you print out the data for func, you'll see that it is the div function that we had defined earlier.

[Video description begins] The following code appears in code cell 36: func(12, 4). [Video description ends]

You can now invoke this div function, pass in input arguments 12 and 4, and you'll get the division result, 12/4= 3. Now that we have access to this handy calc dictionary, you can look up individual functions by key.

[Video description begins] Two lines of code appear in code cell 37. Line 1 is: func = calc_dictionary['*']. Line 2 is: func. The output reads: <function _main_.mul(a, b)>. [Video description ends]

Use the multiplication sign to access the multiplication function, and you can invoke the mul function on say 10, 20 to get the multiplication result which is equal to 200.

[Video description begins] The following code appears in code cell 38: func(10, 20). The output is 200. [Video description ends]

When you look up a function object by key, you don't need to assign that function object to a variable.

[Video description begins] The following code appears in code cell 39: calc_dictionary['*'](100, 3). [Video description ends]

Here, I've looked up the mul function object using the asterisk or multiplication sign. And instead of assigning that to a variable, I directly invoke the function. You know it's a function invocation thanks to the parenthesis and the two input arguments that we've passed in 100, 3. And the result of multiplying these two numbers is 300. This was fun, let's do this again, I won't assign the function object to a variable, instead, I'll directly invoke the plus function that is the add function and add 103. This gives me the result 103.